File: leetcode.com/1760_minimum_limit_of_balls_in_a_bag/solution.py
"""
Binary search on possible result.

TC: O(NlgN)
SC: O(1)
"""
class Solution:
    def minimumSize(self, nums: List[int], maxOperations: int) -> int:

        def is_possible(x):
            t = maxOperations

            for n in nums:
                t -= math.ceil((n - x) / x)

            return t >= 0

        lo, hi = 1, max(nums)
        while lo <= hi:
            mid = (lo + hi) // 2
            if is_possible(mid):
                hi = mid - 1
            else:
                lo = mid + 1

        return lo

File: leetcode.com/1760_minimum_limit_of_balls_in_a_bag/solution.c
int minimumSize(int *nums, int numsSize, int maxOperations)
{
    int lo = 1;
    int hi = 0;

    for (int i = 0 ; i < numsSize ; i++) {
        if (nums[i] > hi)
            hi = nums[i];
    }

    while (lo <= hi) {
        int mid = (lo + hi) / 2;

        // check is possible
        int t = maxOperations;
        for (int i = 0 ; i < numsSize ; i++) {
            t -= ceil((double)(nums[i] - mid) / mid);
        }

        if (t >= 0) {
            hi = mid - 1;
        } else {
            lo = mid + 1;
        }
    }

    return lo;
}
